### PR TITLE
[BC Break] Migrate useGetMany to react-query

### PR DESCRIPTION
**Pull request by fzaninotto**
BC break: useGetMany no longers does query batching by default. For that, users must now use useGetManyAccumulate.

- [x] Create `useGetMany` hook using react-query
- [x] Create `useGetManyAccumulate` hook
- [x] use `useGetManyAccumulate` in `ReferenceField`
- [x] use `useGetManyAccumulate` in `ReferenceArrayField`
- [x] use `useGetManyAccumulate` in `ReferenceInput`
- [x] use `useGetManyAccumulate` in `ReferenceArrayInput`
- [x] use `useGetManyAccumulate` in `ReferenceManyInput`
- [x] Update `useUpdate` so that it optimistically updates the `useGetMany` cache
- [x] Fix tests
- [x] Add upgrade guide

### CODE REVIEW

File: packages/ra-core/src/dataProvider/useGetManyAggregate.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+export const useGetManyAggregate = <RecordType extends Record = Record>(
+    resource: string,
+    params: GetManyParams,
+    options: UseQueryOptions<RecordType[], Error>
+): UseGetManyHookValue<RecordType> => {
+    const dataProvider = useDataProvider();
+    const queryClient = useQueryClient();
+    const { ids } = params;
+
+    return useQuery<RecordType[], Error, RecordType[]>(
```
Comment (fzaninotto) [inline packages/ra-core/src/dataProvider/useGetManyAggregate.ts:98]: this approach is actually more powerful than the previous one: not only do we no longer have to mess with useState, we don't need to batch the side effects anymore either. 

Bonus: both the individual and the aggregated queries are in the cache, so after visiting a post list and displaying a post, its tags will display right away.

**Comment by wattroll** [comment]
Wonderful. And I was just wondering why my `ReferenceField`s aren't going through react-query. I'll give this a try and have a little look at the code :eyes:.

**Comment by wattroll** [comment]
Seem to be causing troubles with `ReferenceField > AutocompleteInput` combination:


In a nutshell: selecting an Autocomplete options item that  isn't in the `useGetMany` cache shows "not found" error message during the fetching of the item and adds an empty string to the autocomplete options:

https://user-images.githubusercontent.com/79251424/146743353-8231be1a-211c-409d-8bea-c0989d92dc59.mp4



Doesn't happen with the items that have been fetched previously:


https://user-images.githubusercontent.com/79251424/146743394-e8e1c5ff-6bc7-44fc-8e02-90938d0d5d5d.mp4

I got this by checking out the branch and running the demo. Hope you can reproduce, if not I'll dig around further.

**Comment by wattroll** [comment]
Can `useGetList` optimistically update the `useGetMany` cache?

**Comment by fzaninotto** [comment]
@wattroll thanks for the report, I'll look into it.

**Review by WiXSL — CHANGES_REQUESTED** [review]
**UPGRADE.md**:

```diff
-Specialized data provider hooks (like `useGetList` and `useUpdate`) have a new signature. There are 2 changes:
+Specialized data provider hooks (like `useGetList`, `useGetMany`, `useGetManyReference`, `useGetOne` and `useUpdate`) have a new signature. There are 2 changes:

File: packages/ra-core/src/dataProvider/useGetMany.ts [hunk 2]
```diff
@@ -1,259 +1,96 @@
...
-                });
-                if (onSuccess) {
-                    onSuccess({ data: emptyArray });
+    params: Partial<GetManyParams> = {},
+    options?: UseQueryOptions<RecordType[], Error>
+): UseGetManyHookValue<RecordType> => {
+    const { ids } = params;
+    const dataProvider = useDataProvider();
+    const queryClient = useQueryClient();
+    return useQuery<RecordType[], Error, RecordType[]>(
```
Comment (WiXSL) [inline packages/ra-core/src/dataProvider/useGetMany.ts:58]:
```suggestion

    return useQuery<RecordType[], Error, RecordType[]>(
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/controller/field/useReferenceArrayFieldController.ts [hunk 3]
```diff
@@ -46,8 +43,8 @@ const defaultSort = { field: null, order: null };
  *
  * @returns {ReferenceArrayProps} The reference props
```
Comment (WiXSL) [inline packages/ra-core/src/controller/field/useReferenceArrayFieldController.ts:44]:
```suggestion
 * @returns {UseReferenceArrayFieldControllerParams} The reference props
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 4]
```diff
@@ -40,7 +38,7 @@ import { ReferenceArrayInputContextValue } from './ReferenceArrayInputContext';
  * @return {Object} controllerProps Fetched data and callbacks for the ReferenceArrayInput components
```
Comment (WiXSL) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts:38]:
```suggestion
 * @return {UseReferenceArrayInputParams} controllerProps Fetched data and callbacks for the ReferenceArrayInput components
```

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 5]
```diff
@@ -40,7 +38,7 @@ import { ReferenceArrayInputContextValue } from './ReferenceArrayInputContext';
...
  */
 export const useReferenceArrayInputController = (
-    props: UseReferenceArrayInputOptions
+    props: UseReferenceArrayInputPArams
```
Comment (WiXSL) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts:41]:
```suggestion
    props: UseReferenceArrayInputParams
```

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 6]
```diff
@@ -360,7 +304,7 @@ const mergeReferences = (ref1: Record[], ref2: Record[]): Record[] => {
     return res;
 };
 
-export interface UseReferenceArrayInputOptions {
+export interface UseReferenceArrayInputPArams {
```
Comment (WiXSL) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts:307]:
```suggestion
export interface UseReferenceArrayInputParams {
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 7]
```diff
@@ -1,10 +1,8 @@
...
-import { useSelector } from 'react-redux';
 import isEqual from 'lodash/isEqual';
-import difference from 'lodash/difference';
 
-import { Record, SortPayload, ReduxState, Identifier } from '../../types';
-import { useGetMany } from '../../dataProvider';
+import { Record, SortPayload, Identifier } from '../../types';
+import { useGetManyAggregate } from '../../dataProvider';
 import { FieldInputProps, useForm } from 'react-final-form';
 import { useGetList } from '../../dataProvider/useGetList';
```
Comment (WiXSL) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts:7]:
```suggestion
```

File: packages/ra-core/src/controller/input/useReferenceArrayInputController.ts [hunk 8]
```diff
@@ -1,10 +1,8 @@
 import { useMemo, useState, useEffect, useRef, useCallback } from 'react';
-import { useSelector } from 'react-redux';
 import isEqual from 'lodash/isEqual';
-import difference from 'lodash/difference';
 
-import { Record, SortPayload, ReduxState, Identifier } from '../../types';
-import { useGetMany } from '../../dataProvider';
+import { Record, SortPayload, Identifier } from '../../types';
+import { useGetManyAggregate } from '../../dataProvider';
```
Comment (WiXSL) [inline packages/ra-core/src/controller/input/useReferenceArrayInputController.ts:5]:
```suggestion
import { useGetManyAggregate, useGetList } from '../../dataProvider';
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — COMMENTED** [review]
`useGetManyAggregate` hook is not mentioned at all in the docs. Maybe in another PR ?

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — COMMENTED** [review]
The first time I filter the `Orders` list o the demo by `Customers` I get this error:

**Comment by fzaninotto** [comment]
> `useGetManyAggregate` hook is not mentioned at all in tehe docs. Maybe in another PR ?

I'm not sure I want to make it public API yet.

File: UPGRADE.md [hunk 9]
```diff
@@ -110,7 +110,7 @@ This should be mostly transparent for you unless:
 
 ## Changed Signature Of Data Provider Hooks
 
-Specialized data provider hooks (like `useGetList` and `useUpdate`) have a new signature. There are 2 changes:
+Specialized data provider hooks (like `useGetOne`, `useGetList`, `useGetMany` and `useUpdate`) have a new signature. There are 2 changes:
```
Comment (WiXSL) [inline UPGRADE.md:113]:
```suggestion
Specialized data provider hooks (like `useGetList`, `useGetMany`, `useGetManyReference`, `useGetOne` and `useUpdate`) have a new signature. There are 2 changes:
```

Comment (WiXSL) [reply UPGRADE.md:113]: I've added `useGetManyReference` here, is simpler than reviewing the same in the other PR

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Comment by fzaninotto** [comment]
Thanks a lot for the review, @WiXSL!

**Pull request merged**